crates/bls12381/src/fields/fp.rs

```diff
@@ -0,0 +1,504 @@
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fp<F> = EmulatedFieldElement<F, Bls12381FpParams>;
+
+#[derive(Clone)]
+pub struct AllocatedFieldElement<F: PrimeField + PrimeFieldBits>(pub(crate) Bls12381Fp<F>);
+
+#[derive(Debug)]
+pub struct Bls12381FrParams;
+
+impl EmulatedFieldParams for Bls12381FrParams {
+    fn num_limbs() -> usize {
+        4
+    }
+
+    fn bits_per_limb() -> usize {
+        64
+    }
+
+    fn modulus() -> BigInt {
+        BigInt::parse_bytes(
+            b"73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001",
+            16,
+        )
+        .unwrap()
+    }
+
+    fn is_modulus_pseudo_mersenne() -> bool {
+        false
+    }
+
+    fn pseudo_mersenne_params() -> Option<PseudoMersennePrime> {
+        None
+    }
+}
+
+pub type Bls12381Fr<F> = EmulatedFieldElement<F, Bls12381FrParams>;
+
+impl<F> From<&BlsFp> for AllocatedFieldElement<F>
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    fn from(value: &BlsFp) -> Self {
+        let bytes = value.to_bytes();
+        assert!(bytes.len() == 48);
+        let val = BigInt::from_bytes_be(Sign::Plus, &bytes);
+        Self(Bls12381Fp::<F>::from(&val))
+    }
+}
+
+pub(crate) fn bigint_to_fpelem(val: &BigInt) -> Option<BlsFp> {
+    use num_traits::Zero;
+    if val >= &Bls12381FpParams::modulus() {
+        return None;
+    }
+    let be_bytes = BigInt::to_bytes_be(val);
+    if be_bytes.0 != Sign::Plus {
+        // the sign is only non-positive if the value is exactly 0
+        assert!(val == &BigInt::zero());
+    }
+    let mut bytes: Vec<u8> = be_bytes.1;
+    assert!(bytes.len() <= 48);
+    while bytes.len() < 48 {
+        bytes.insert(0, 0);
+    } // TODO: here we pad with 0 bytes up to 48 bytes, but maybe something better could be done instead
+    let bytes: [u8; 48] = bytes.try_into().unwrap();
+    Some(BlsFp::from_bytes(&bytes).unwrap())
+}
+
+pub(crate) fn emulated_to_native<F>(value: &Bls12381Fp<F>) -> BlsFp
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    use std::ops::Rem;
+    let p = &Bls12381FpParams::modulus();
+    let val = BigInt::from(value).rem(p);
+    bigint_to_fpelem(&val).unwrap()
+}
+
+impl<F> From<&AllocatedFieldElement<F>> for BlsFp
+where
+    F: PrimeField + PrimeFieldBits,
+{
+    fn from(value: &AllocatedFieldElement<F>) -> Self {
+        emulated_to_native(&value.0)
+    }
+}
+
+impl<F: PrimeField + PrimeFieldBits> AllocatedFieldElement<F> {
+    pub fn from_dec(val: &str) -> Result<Self, SynthesisError> {
+        let bigint =
+            BigInt::parse_bytes(val.as_bytes(), 10).ok_or(SynthesisError::Unsatisfiable)?;
+        let elm = bigint_to_fpelem(&bigint).ok_or(SynthesisError::Unsatisfiable)?;
+        Ok(Self::from(&elm))
+    }
```

Comment:
It seems like a horrible abuse of error types to use a `SynthesisError` when you are not synthesizing — and `Unsatisfiable` doesn't seem applicable either. I think you need a different error/type here.

Reply:
I will make this function return an `Option<Self>` instead, this function will really only be called with constant strings and should never fail in practice. The caller can either `.unwrap()` (in the constant case) or do an `ok_or` if necessary.